src/lin.py

```diff
@@ -1,6 +1,50 @@
 """Implementation of a linear time exact matching algorithm."""
 
 import argparse
+import fasta
+import fastq
+
+def border_array(x: str) -> list[int]:
+    """
+    Construct the border array for x.
+
+    >>> border_array("aaba")
+    [0, 1, 0, 1]
+    >>> border_array("ississippi")
+    [0, 0, 0, 1, 2, 3, 4, 0, 0, 1]
+    >>> border_array("")
+    []
+    """
+    ba = [0] * len(x)
+    for i in range(len(x)):
+        if i == 0:
+            continue
+        if x[ba[i-1]] == x[i]:
+            ba[i] = ba[i-1] + 1
```

Comment:
Two things here:

1) You are correctly extending the previous border if you can, but if you can't, you will reset the border to zero (implicitly, because `ba` is initialised to zeros). You get correct border lengths, so that will not be a problem, but you do not get the longest borders; if the first can't be extended, the next choice is not the empty string but the second-longest or third-longest, etc. By not looking at the sequence of shorter borders you are hurting the running time.

2) This is a minor thing, but if you want to skip index 0 you might as well use `range(1, len(x))` instead of the `continue`.